### PR TITLE
Added `artifacts list` feature

### PR DESCRIPTION
**Pull request by siredmar**
This feature adds the ability to list artifacts. There is a detail level argument. 

image

```
/mender-cli artifacts --help     
Operations on mender artifacts.

Usage:
  mender-cli artifacts [command]

Available Commands:
  list        List mender artifacts from the Mender server.
  upload      Upload mender artifact to the Mender server.
...

./mender-cli artifacts list --help
List mender artifacts from the Mender server.

Usage:
  mender-cli artifacts list [flags]

Flags:
  -d, --detail int   artifacts list detail level [0..2]
  -h, --help         help for list
...
```

**Detail level 0 (default):**

```
/mender-cli artifacts list                  
ID: 26615230-3aa5-40f4-8f88-1922bbf64b91
Name: mender-demo-artifact-2.5.0
Signed: false
Modfied: 2020-10-09 13:30:36.987 +0000 UTC
Size: 2034176
Description: For demo purposes. This Artifact runs simple webserver on your device serving a static web page that you can view in the browser.
--------------------------------------------------------------------------------
ID: 4f38e59f-b10c-460b-86c1-a133c324804e
Name: release-2
Signed: false
Modfied: 2020-10-13 13:19:28.065 +0000 UTC
Size: 56369664
Description: Release 1
--------------------------------------------------------------------------------
ID: 2113e5aa-704e-4ce0-b942-8cfe99b7fb0b
Name: release-3
Signed: false
Modfied: 2020-10-14 11:14:12.022 +0000 UTC
Size: 56370176
Description: 
--------------------------------------------------------------------------------
```

**Detail level 1:**

```
./mender-cli artifacts list -d 1   
ID: 26615230-3aa5-40f4-8f88-1922bbf64b91
Name: mender-demo-artifact-2.5.0
Signed: false
Modfied: 2020-10-09 13:30:36.987 +0000 UTC
Size: 2034176
Description: For demo purposes. This Artifact runs simple webserver on your device serving a static web page that you can view in the browser.
Compatible device types:
  beaglebone
  beaglebone-yocto
  beaglebone-yocto-grub
  generic-armv6
  generic-x86_64
  qemux86-64
  raspberrypi0w
  raspberrypi0-wifi
  raspberrypi3
  raspberrypi4
  raspberrypi
Artifact format: mender
Format version: 3
--------------------------------------------------------------------------------
ID: 4f38e59f-b10c-460b-86c1-a133c324804e
Name: release-2
Signed: false
Modfied: 2020-10-13 13:19:28.065 +0000 UTC
Size: 56369664
Description: Release 1
Compatible device types:
  raspberrypi4
Artifact format: mender
Format version: 3
--------------------------------------------------------------------------------
ID: 2113e5aa-704e-4ce0-b942-8cfe99b7fb0b
Name: release-3
Signed: false
Modfied: 2020-10-14 11:14:12.022 +0000 UTC
Size: 56370176
Description: 
Compatible device types:
  raspberrypi4
Artifact format: mender
Format version: 3
--------------------------------------------------------------------------------
```

**Detail level 2:**

```
./mender-cli artifacts list -d 2   
ID: 26615230-3aa5-40f4-8f88-1922bbf64b91
Name: mender-demo-artifact-2.5.0
Signed: false
Modfied: 2020-10-09 13:30:36.987 +0000 UTC
Size: 2034176
Description: For demo purposes. This Artifact runs simple webserver on your device serving a static web page that you can view in the browser.
Artifact provides: mender-demo-artifact-2.5.0
Artifact depends:
  beaglebone
  beaglebone-yocto
  beaglebone-yocto-grub
  generic-armv6
  generic-x86_64
  qemux86-64
  raspberrypi0w
  raspberrypi0-wifi
  raspberrypi3
  raspberrypi4
  raspberrypi
Updates:
  Type: directory
  Files:
    Name: update.tar
    Checksum: 78956caaa6a622f925a55fd03271379f0ef6d79521962c5192d2aba72a3ccc4c
    Size: 3235840
    Date: 2020-10-06 11:06:24 +0000 UTC

    Name: dest_dir
    Checksum: 15c9b27e96852e92dddade6417561a6b6ed701efa179063c5a501c3d435ae20b
    Size: 19
    Date: 2020-10-06 11:06:24 +0000 UTC

--------------------------------------------------------------------------------
ID: 4f38e59f-b10c-460b-86c1-a133c324804e
Name: release-2
Signed: false
Modfied: 2020-10-13 13:19:28.065 +0000 UTC
Size: 56369664
Description: Release 1
Artifact provides: release-2
Artifact depends:
  raspberrypi4
Updates:
  Type: rootfs-image
  Files:
    Name: core-image-base-raspberrypi4.ext4
    Checksum: fc597936456258560e44ef3a0f0520121b07b122315dc2e3cb0742542ccb3962
    Size: 209715200
    Date: 2020-10-06 11:55:35 +0000 UTC
--------------------------------------------------------------------------------
ID: 2113e5aa-704e-4ce0-b942-8cfe99b7fb0b
Name: release-3
Signed: false
Modfied: 2020-10-14 11:14:12.022 +0000 UTC
Size: 56370176
Description: 
Artifact provides: release-3
Artifact depends:
  raspberrypi4
Updates:
  Type: rootfs-image
  Files:
    Name: core-image-base-raspberrypi4.ext4
    Checksum: 036de4ed80653e4646303e3a5a6e1609014029f49b392cabcaa2d45cbb54b0ff
    Size: 209715200
    Date: 2020-10-07 14:14:49 +0000 UTC
--------------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by oleorhagen** [comment]
@siredmar do you want me to hold off on the review?

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:100]: Question:
I don't think this captures this error completely. I think that the cached token can be present, but you still get `Unathorized`. Which would lead to the HTTP request returning `401` below. So I think an explicit login is required, as compared to this?

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:107]: Suggestion:
```suggestion
		return errors.Wrap(err, "Failed to create HTTP request")
```

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:112]:
```suggestion
	log.Verbf("sending request: \n%s", string(reqDump))
```

**Comment by siredmar** [comment]
@olehermanse I rebased this to master after #76 was merged. So you can go ahead with the review.

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:116]: Suggestion:
Also handle the HTTP error codes returned. As this is not an error. The request in itself is succesful, therefore the HTTP status code needs to be checked here, otherwise there will be no data to unmarshal in the case of HTTP error codes with no json in the body response.

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:138]:
```suggestion
	fmt.Printf("ID: %s", a.ID)
```

Comment (oleorhagen) [reply client/deployments/client.go:138]: And the same below :+1:

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:200]: :+1:

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:163]: Suggestion:
```suggestion
				fmt.Println(fmt.Sprintf("\tName: %s", f.Name))
```

File: cmd/artifacts_list.go
Comment (oleorhagen) [inline cmd/artifacts_list.go:28]:
```suggestion
	Short: "List Mender Artifacts from the Mender server.",
```

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:137]: This could potentially be a nice use-case for Go's text template package: https://golang.org/pkg/text%2Ftemplate/

But this is no requirement, this is just fine. But if you're feeling adventurous :)

File: client/deployments/client.go
Comment (oleorhagen) [reply client/deployments/client.go:112]: Also interesting that it has `Verbf`, but not `Infof`

**Comment by siredmar** [comment]
@oleorhagen gitlab checks are failing with this message:
```
echo "-- checking cyclometric complexity > 15"
-- checking cyclometric complexity > 15
gocyclo -over 15 ./main.go ./log/log.go ./client/useradm/client.go ./client/deployments/client.go ./cmd/artifacts_list.go ./cmd/artifact_upload.go ./cmd/artifacts.go ./cmd/root.go ./cmd/util.go ./cmd/login.go
make: gocyclo: Command not found
```

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:111]: Question:
The `httputil.DumpRequest` documentation gives me this:
```
 If DumpRequest returns an error, the state of req is undefined
```
Which worries me a little. Maybe we should check the error here?

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:40]: Real nice

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:74]:
```suggestion
	artifactsListURL  = artifactUploadURL
```

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:144]: Question:
Why do we not want to list this at detail-level `2`?

**Review by oleorhagen — COMMENTED** [review]
Real nice!

I've left some comments and suggestions, but overall this looks real good. I like the idea of this command, and the output of it. Handy addition!

**Comment by oleorhagen** [comment]
> @oleorhagen gitlab checks are failing with this message:
> 
> ```
> echo "-- checking cyclometric complexity > 15"
> -- checking cyclometric complexity > 15
> gocyclo -over 15 ./main.go ./log/log.go ./client/useradm/client.go ./client/deployments/client.go ./cmd/artifacts_list.go ./cmd/artifact_upload.go ./cmd/artifacts.go ./cmd/root.go ./cmd/util.go ./cmd/login.go
> make: gocyclo: Command not found
> ```

Yes, this is not on you. I will look into it :)

File: client/deployments/client.go
Comment (oleorhagen) [inline client/deployments/client.go:60]: Also, you are parsing the `MetaData`, but never printing it

Comment (oleorhagen) [reply client/deployments/client.go:60]: Same for the `Signed` field I think.

Comment (siredmar) [reply client/deployments/client.go:60]: Could you tell more more about MetaData? `meta_data is an object of unknown structure as this is dependent of update type (also custom defined by user)`
I would introduce a detail level 4 to enable printing this `meta_data`. 

`Signed` is printed here https://github.com/siredmar/mender-cli/blob/5ea07ab72cb86e1e7220d7799dc9de5dbd3072b4/client/deployments/client.go#L140

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:144]: I'm fine with that. will move them to level 2

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:74]: OK

File: client/deployments/client.go
Comment (oleorhagen) [reply client/deployments/client.go:60]: Sorry, missed the Signed part (sausage fingers).

The `MetaData` is just a generic JSON structure, with not requirements on the structure, as far as Mender is concerned. Hence, this can be whatever, as long as it is JSON. I guess this can be printed in `level 4` as  you suggest, with something along the lines of:
```
fmt.Printf("%+v\n", metaData)
```
or something simlar like: `%#v` as the verb maybe. Experiment a little, and see which you prefer :)

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:100]: In fact i like the idea of having a file containing the token for auth lying around somewhere. I already thought of a way to put the token in the `.mender-clirc` file instead of putting user and pw in it. 
But to make this feature really useful there must be a way to create tokens in the cloud app with specific rights and a way to revoke those tokens. So basically i'd preferr trying to read the token from a file (for the moment from cached file, later from `.mender-clirc`). Currently i'd stick with the solution and plan to implement the feature mentioned above. 

With this implementation when the token is invalid an error is reported:

```
./mender-cli artifacts list
Using configuration file: /home/armin/.mender-clirc
FAILURE: Get /deployments/artifacts request failed: Get "https://hosted.mender.io/api/management/v1/deployments/artifacts": net/http: invalid header field value "Bearer eyJhbGciOiJSUzI1.........RKlU\n" for key Authorization
```

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:107]: OK

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:112]: OK

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:138]: Usign fmt.Printf.

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:163]: ok`

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:200]: ok

File: cmd/artifacts_list.go
Comment (siredmar) [reply cmd/artifacts_list.go:28]: ok

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:137]: thx for the hint. Will look into it, but leaving this for now the way it is.

File: client/deployments/client.go
Comment (oleorhagen) [reply client/deployments/client.go:100]: > In fact i like the idea of having a file containing the token for auth lying around somewhere. I already thought of a way to put the token in the `.mender-clirc` file instead of putting user and pw in it.

Yes, I agree.

> But to make this feature really useful there must be a way to create tokens in the cloud app with specific rights and a way to revoke those tokens. So basically i'd preferr trying to read the token from a file (for the moment from cached file, later from `.mender-clirc`). Currently i'd stick with the solution and plan to implement the feature mentioned above.
> 
> With this implementation when the token is invalid an error is reported:
> 
> ```
> ./mender-cli artifacts list
> Using configuration file: /home/armin/.mender-clirc
> FAILURE: Get /deployments/artifacts request failed: Get "https://hosted.mender.io/api/management/v1/deployments/artifacts": net/http: invalid header field value "Bearer eyJhbGciOiJSUzI1.........RKlU\n" for key Authorization
> ```

This I'm not sure of. Should this not return `401` ?

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:111]: returning err if != nil

Comment (oleorhagen) [reply client/deployments/client.go:111]: > returning err if != nil

Yes

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:60]: sticking with `fmt.Printf("  MetaData: %v\n", v.MetaData)`

**Comment by oleorhagen** [comment]
> > @oleorhagen gitlab checks are failing with this message:
> > ```
> > echo "-- checking cyclometric complexity > 15"
> > -- checking cyclometric complexity > 15
> > gocyclo -over 15 ./main.go ./log/log.go ./client/useradm/client.go ./client/deployments/client.go ./cmd/artifacts_list.go ./cmd/artifact_upload.go ./cmd/artifacts.go ./cmd/root.go ./cmd/util.go ./cmd/login.go
> > make: gocyclo: Command not found
> > ```
> 
> Yes, this is not on you. I will look into it :)

Some of ur commits are missing a changelog tag. 

So just add them, and then rebase on top of master and force push and this should be good :)

**Comment by siredmar** [comment]
> > > @oleorhagen gitlab checks are failing with this message:
> > > ```
> > > echo "-- checking cyclometric complexity > 15"
> > > -- checking cyclometric complexity > 15
> > > gocyclo -over 15 ./main.go ./log/log.go ./client/useradm/client.go ./client/deployments/client.go ./cmd/artifacts_list.go ./cmd/artifact_upload.go ./cmd/artifacts.go ./cmd/root.go ./cmd/util.go ./cmd/login.go
> > > make: gocyclo: Command not found
> > > ```
> > 
> > 
> > Yes, this is not on you. I will look into it :)
> 
> Some of ur commits are missing a changelog tag.
> 
> So just add them, and then rebase on top of master and force push and this should be good :)

Dang it! Thx :)

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:103]: Yes, you are right. I modified the authtoken file manually, Unfortunately this is not the way to go to invalidate the token... Anyway, is it okay for you to stick with the current solution for now and build the more advanced token handling in the `.mender-clirc` file?

Comment (oleorhagen) [reply client/deployments/client.go:103]: Yes, that is fine with me for now :)

File: client/deployments/client.go
Comment (oleorhagen) [reply client/deployments/client.go:122]: @siredmar what are your thoughts on this one?

**Comment by siredmar** [comment]
@oleorhagen nothing left to do here for me.

**Comment by oleorhagen** [comment]
> @oleorhagen nothing left to do here for me.

https://github.com/mendersoftware/mender-cli/pull/77#discussion_r507716451

Just curious about your thoughts on this one first.

File: client/deployments/client.go
Comment (siredmar) [reply client/deployments/client.go:122]: Oh i missed that one. Sure, check != 200?

Comment (oleorhagen) [reply client/deployments/client.go:122]: https://docs.mender.io/api/#list-releases-responses

Yup, looks like that should do the trick for now :)

**Review by oleorhagen — APPROVED** [review]
Nice!

**Comment by oleorhagen** [comment]
I'll merge this as soon as the pipes are :green_circle:

**Comment by oleorhagen** [comment]
@siredmar and also remember the `--signoff` with your commits :)

**Comment by siredmar** [comment]
> @siredmar and also remember the `--signoff` with your commits :)

🙄 this again :)

**Comment by oleorhagen** [comment]
> > @siredmar and also remember the `--signoff` with your commits :)
> 
> roll_eyes this again :)

Yes, sry. Formalities... :)

I just add it to my gitconfig, so that I always signoff

**Comment by oleorhagen** [comment]
Well, done! These are making the tool a lot nicer imo

**Pull request merged**
